inc/saiswitch.h

```diff
@@ -325,6 +325,135 @@ typedef enum _sai_switch_failover_config_mode_t
 
 } sai_switch_failover_config_mode_t;
 
+/**
+ * @brief Defines tunnel type
+ */
+typedef enum _sai_tunnel_type_t
+{
+    SAI_TUNNEL_TYPE_IPINIP,
+
+    SAI_TUNNEL_TYPE_IPINIP_GRE,
+
+    SAI_TUNNEL_TYPE_VXLAN,
+
+    SAI_TUNNEL_TYPE_MPLS,
+
+} sai_tunnel_type_t;
+
+/**
+ * @brief Defines tunnel encap ECN mode
+ */
+typedef enum _sai_tunnel_encap_ecn_mode_t
+{
+    /**
+     * @brief Normal mode behavior defined in RFC 6040
+     * section 4.1 copy from inner
+     */
+    SAI_TUNNEL_ENCAP_ECN_MODE_STANDARD,
+
+    /**
+     * @brief User defined behavior.
+     */
+    SAI_TUNNEL_ENCAP_ECN_MODE_USER_DEFINED
+
+} sai_tunnel_encap_ecn_mode_t;
+
+/**
+ * @brief Defines tunnel decap ECN mode
+ */
+typedef enum _sai_tunnel_decap_ecn_mode_t
+{
+    /**
+     * @brief Behavior defined in RFC 6040 section 4.2
+     */
+    SAI_TUNNEL_DECAP_ECN_MODE_STANDARD,
+
+    /**
+     * @brief Copy from outer ECN
+     */
+    SAI_TUNNEL_DECAP_ECN_MODE_COPY_FROM_OUTER,
+
+    /**
+     * @brief User defined behavior
+     */
+    SAI_TUNNEL_DECAP_ECN_MODE_USER_DEFINED
+
+} sai_tunnel_decap_ecn_mode_t;
+
+/**
+ * @brief Defines tunnel attributes at switch level
+ */
+typedef enum _sai_switch_tunnel_attr_t
+{
+    /**
+     * @brief Start of attributes
+     */
+    SAI_SWITCH_TUNNEL_ATTR_START,
+
+    /**
+     * @brief Tunnel type
+     *
+     * @type sai_tunnel_type_t
+     * @flags MANDATORY_ON_CREATE | CREATE_ONLY
```

Comment:
should this also be marked as KEY ?
as we allow up to one object instance per switch ID per tunnel ID, meaning there can't be 2 objects for example for same switch ID for same tunnel type

Reply:
Done
https://github.com/opencomputeproject/SAI/pull/1173/commits/1ed0f2e544818c598b455b3046e159689e5fbfa4